### PR TITLE
Remove OpenhabCloud's Authorization Header to OH3

### PR DESCRIPTION
**Pull request by lsafelix75**
The problem that i am facing is openhabcloud actually passes its credentials via authorization header to remote openhab instance, which username and passwords are different.

### CODE REVIEW

**Comment by digitaldan** [comment]
Hi , thanks for the PR.  Can you explain your use case a little more?  Are you trying to use BASIC auth vs the token based auth in OH3?  Can you explain what's not working (as its working for others right now). Are you using the public myopenhab.org service or are you hosting your own ?

**Comment by lsafelix75** [comment]
Hi

This is self hosted openhabcloud. I believe some people are facing the same issue (in my case I use chrome) that when we click on home.Xxxx.Xxxx we are greeted with a pop up asking for username and password. This username and password are passed down to OH3 as Basic Auth  and render invalid username and password which causing browser screen blacken.

**Comment by digitaldan** [comment]
Hi , so this is working on myopenhab.org, so i'm trying to understand why its not working for you before we change things.  When you click on that link, it should be using the already set cookie to authorize, not BASIC auth, so it may be you do not have the `subDomainCookies` flag turned on in your config?

**Comment by lsafelix75** [comment]
> Hi , so this is working on myopenhab.org, so i'm trying to understand why its not working for you before we change things. When you click on that link, it should be using the already set cookie to authorize, not BASIC auth, so it may be you do not have the `subDomainCookies` flag turned on in your config?


Yes, i have tried that but the browser still asking me for username and password. See headers below when i click on home.openhabcloud.local:82:

GET / HTTP/1.1
Host: home.openhabcloud.local:82
Connection: keep-alive
Cache-Control: max-age=0
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.150 Safari/537.36 Edg/88.0.705.63
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9
Referer: http://openhabcloud.local:82/
Accept-Encoding: gzip, deflate
Accept-Language: en-GB,en;q=0.9,en-US;q=0.8,zh-TW;q=0.7,zh;q=0.6
Cookie: __utma=76538187.1263348583.1613206500.1613206500.1613206500.1; __utmc=76538187; __utmz=76538187.1613206500.1.1.utmcsr=(direct)|utmccn=(direct)|utmcmd=(none); __utma=53141566.919369639.1613206539.1613206539.1613206539.1; __utmc=53141566; __utmz=53141566.1613206539.1.1.utmcsr=openhabcloud.local:82|utmccn=(referral)|utmcmd=referral|utmcct=/; X-OPENHAB-AUTH-HEADER=true; X-OPENHAB-SESSIONID=1e2fd27c-433e-4b71-9afc-5c147a6dbf3f; X-OPENHAB-SESSIONID=f196f8ba-92d2-4359-9ee0-10d032ca0bd4; connect.sid=s%3Al_ej1PwfGyh7gHY3e1or6X5mIXmdcsZ1.uDW5d2afUBc%2BNnfJp3DUH3q3uEvtf%2FM9GjeE2Gj9hWA; connect.sid=s%3A9-1jXaFU1R468O2Slhls8l7RqGfCAq6k.DEKUI7Wcx4HG1SjKRswwtplWCMbrv%2F4kY87KwGXIvpM

**Comment by digitaldan** [comment]
its hard to help with out a more detailed setup of how you are running this.  What your config file looks like, what hosts and ports you are using, what load balancer you are using, what its config looks like , etc......

**Comment by lsafelix75** [comment]
it is a simple 1 server system, and following is openhabcloud config.json:
```

{
    "system": {
        "host": "openhabcloud.local",
        "port": "82",
        "protocol": "http",
	      "logger" : {
            "dir": "./logs",
            "maxFiles" : "7d",
            "level" : "debug",
            "morganOption" : "dev"
        },
        "proxyHost": "home.openhabcloud.local",
        "proxyPort": "82",
        "subDomainCookies": false,
        "muteNotifications": false

    },
    "express":{
      "key" : "some express key"
    },
    "mongodb": {
        "hosts": ["127.0.0.1:27017"],
        "db": "openhab"
    },
    "redis": {
        "host": "127.0.0.1",
        "port": "6379",
        "password": "secret"
    },
    "mailer": {
        "host" : "somehost",
        "port": 465,
        "secureConnection": true,
        "user": "someuser",
        "password": "somepass",
        "from": "<your-mail>"
    },
    "gcm" : {
      "jid": "something@gcm.googleapis.com",
      "password": "password"
    },
    "ifttt" : {
      "iftttChannelKey" : "key",
      "iftttTestToken" : "token"
    },
    "legal": {
    	"terms" : "",
        "policy": ""
    },
     "registration_enabled": false
}

```

This is nginx config:
```

# You may add here your
# server {
#	...
# }
# statements for each of your virtual hosts to this file

#socket.io connections (from OH) NEED ip_hash routing, all other connections will be cookie or round robin
upstream socketapp {
        server 127.0.0.1:3000;
}

upstream socketapp2 {
        server 127.0.0.1:3001;
}

upstream socketapp3 {
        server 127.0.0.1:8083;
}

#Connections that need to be proxied, need same entries as "socketapp"
upstream proxyapp {
        server 127.0.0.1:3000;
}

#all other connections to serve openHAB-Cloud requests (not proxied)
upstream webapp {
        server 127.0.0.1:3000;
}



server {
    listen 82;
    
    access_log /var/log/nginx/thecode-cloud-access.log;  
    error_log /var/log/nginx/thecode-cloud-error.log debug; 
    server_name openhabcloud.local;
 
    root /var/www/openhab-cloud;
    index index.html index.htm;
   
    location ~ /.well-known {
        allow all;
    }

    location /css {
        alias  /var/www/thecode-cloud/public/css;
    }
    location /js {
        alias /var/www/thecode-cloud/public/js;
    }
    location /img {
        alias /var/www/thecode-cloud/public/img;
    }
    location /bootstrap {
        alias /var/www/thecode-cloud/public/bootstrap;
    }
    location /font-icons {
        alias /var/www/thecode-cloud/public/font-icons;
    }
    location /fonts {
        alias /var/www/thecode-cloud/public/fonts;
    }
    location /js-plugin {
        alias /var/www/thecode-cloud/public/js-plugin;
    }
    location /downloads {
        alias /var/www/thecode-cloud/public/downloads;
    }


    #OH Socket.io Connections (from OH)
    location /socket.io{
        proxy_pass http://socketapp;
        proxy_redirect off;
        proxy_http_version 1.1;
        proxy_set_header Host $http_host;
        proxy_set_header X-NginX-Proxy true;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        #proxy_set_header X-Forwarded-Proto https;
        proxy_read_timeout 86400;

    }

   include /etc/nginx/sites-enabled/redirects.config;

   location ~ ^/(rest|images|static|rrdchart.png|chart|openhab.app|WebApp|CMD|cometVisu|proxy|greent|jquery|classicui|ui|basicui|doc|start|icon|habmin|remote|habpanel|ifttt/v1/actions/command){
        set $upstream_server proxyapp;
        #if we have a cookie, try using this server
        #also, ifIsEvil https://www.nginx.com/resources/wiki/start/topics/depth/ifisevil/
        #for why we use this as little as possible
        if ($http_cookie ~ "CloudServer=(\S+)\%3A(\d+).*") {
            set $upstream_host $1;
            set $upstream_port $2;
            set $upstream_server "${upstream_host}:${upstream_port}";
        }
        proxy_pass http://$upstream_server;
        proxy_redirect off;
        proxy_http_version 1.1;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr ;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for ;
        proxy_set_header X-Forwarded-Proto https;
                proxy_intercept_errors on;
                recursive_error_pages on;
        error_page 307 = @handle_proxy;
        #if this server is down we need to try a new upstream
        error_page 500 501 502 503 504 = @proxy_down;
   }


 #All other non proxy connections
    location / {
         proxy_pass http://webapp;
         proxy_redirect off;
         proxy_http_version 1.1;
         proxy_set_header Host $host;
         proxy_set_header X-Real-IP $remote_addr ;
         proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for ;
         proxy_set_header X-Forwarded-Proto http;
    }

}

server {
    #ssl_certificate /opt/thecode/certs/gateway_cert.pem;
    #ssl_certificate_key /opt/thecode/certs/gateway_key.key;

    
    listen 82;
    server_name home.openhabcloud.local;

    charset utf-8; 

    location ~ /.well-known {
      allow all;
    }

     include /etc/nginx/sites-enabled/redirects.config;
 
     location / {
		set $upstream_server proxyapp;
		#if we have a cookie, try using this server
		#also, ifIsEvil https://www.nginx.com/resources/wiki/start/topics/depth/ifisevil/
		#for why we use this as little as possible
		if ($http_cookie ~ "CloudServer=(\S+)\%3A(\d+).*") {
			set $upstream_host $1;
			set $upstream_port $2;
			set $upstream_server "${upstream_host}:${upstream_port}";
		}
		proxy_pass http://$upstream_server;
		proxy_redirect off;
		proxy_http_version 1.1;
		proxy_set_header Host $host;
		proxy_set_header X-Real-IP $remote_addr ;
		proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for ;
		proxy_set_header X-Forwarded-Proto http;
		proxy_intercept_errors on;
		recursive_error_pages on;
		error_page 307 = @handle_proxy;
		#if this server is down we need to try a new upstream
		error_page 500 501 502 503 504 = @proxy_down;
	}
}

```

File: routes/index.js [hunk 1]
```diff
@@ -370,13 +370,16 @@ Routes.prototype.proxyRouteOpenhab = function (req, res) {
     delete requestHeaders['x-forwarded-for'];
     delete requestHeaders['x-forwarded-proto'];
     delete requestHeaders['connection'];
-    requestHeaders['host'] = req.headers.host || system.getHost() + ':' + system.getPort();
+    requestHeaders['host'] = req.headers["x-forwarded-host"] || system.getHost() + ':' + system.getPort();
```
Comment (digitaldan) [inline routes/index.js:373]: if you notice on line 384 the code resets the host value, so this change is not doing anything when we are doing a full proxy through the "home" domain.  I understand why you changed this, but did you test it out to ensure it worked?   I agree we should be setting the forwarded host ip,  i don't know exactly why the author did it this way.

**Comment by lsafelix75** [comment]
It didn't work with Remoteopenhab plugin if standard ports are not configured. As openhab remote instance uses  this value to construct uris for its responses.

File: routes/index.js [hunk 2]
```diff
@@ -370,13 +370,16 @@ Routes.prototype.proxyRouteOpenhab = function (req, res) {
...
     requestHeaders['user-agent'] = 'openhab-cloud/0.0.1';
     // Strip off path prefix for remote vhosts hack
     var requestPath = req.path;
     if (requestPath.indexOf('/remote/') === 0) {
         requestPath = requestPath.replace('/remote', '');
         // TODO: this is too dirty :-(
+        if(requestPath.indexOf('/remote/auth') !== 0){
+            delete requestHeaders['authorization'];
```
Comment (digitaldan) [inline routes/index.js:381]: Just an FYI, the openHAB 3 ui will automatically use another header for authentication `X-OPENHAB-AUTH-HEADER`

**Comment by lsafelix75** [comment]
OK. Is this a new PR? So far I'm not able to get it through.

**Comment by digitaldan** [comment]
if you load up your chrome dev tools for your instance can you see what the cookie domain is?  In the dev tools this is under `Application` -> `Storage` -> `Cookies` -> `http://openhabcloud.local` ?  For example, in myopenhab this looks like:
image

Your domain should be prefixed with a `.` , which means it can be shared across hosts in that domain.  
I also see you are not running this with SSL?  I don't think that should be an issue, but trying to chase down what is different about your setup vs myopenhab and others self hosted instances.

**Comment by digitaldan** [comment]
>OK. Is this a new PR? So far I'm not able to get it through.

We added this quite some time ago when we were validating that myopenHAB would work with OH3 as that was a requirement for very first milestone release.

File: routes/index.js (hunk 2 above)
Comment (lsafelix75) [reply routes/index.js:381]: apparently this is no needed. After i reboot the server that is hosting openhab-cloud instance. Everything is working as intended. Thanks for your time.

**Pull request closed without merging**